datalad/support/gitrepo.py

```diff
@@ -3123,85 +3170,137 @@ def _diffstatus_get_state_props(self, f, from_state, to_state,
-                # identical id, so we need to check whether the file is
-                # gone before declaring it clean. This working tree
-                # check is irrelevant and wrong if to is a ref.
-                state = 'clean' \
-                    if against_commit or (f.exists() or f.is_symlink()) \
-                    else 'deleted'
-        else:
-            # change in git record, or on disk
-            # for subdatasets leave the 'modified' judgement to the caller
-            # for supporting corner cases, such as adjusted branch
-            # which require inspection of a subdataset
-            # TODO we could have a new file that is already staged
-            # but had subsequent modifications done to it that are
-            # unstaged. Such file would presently show up as 'added'
-            # ATM I think this is OK, but worth stating...
-            state = ('modified'
-                     if against_commit or to_state['type'] != 'dataset'
-                     else None
-                    ) if f.exists() or f.is_symlink() else 'deleted'
-            # TODO record before and after state for diff-like use
-            # cases
+        state = self._diffstatus_get_state(
+            f, from_state, to_state,
+            modified_in_worktree,
+            against_commit, eval_submodule_state,
+        )
+
+        # compile properties of the diff state
+        # TODO sort out
+        props = {}
+
+        to_sha = to_state.get('gitshasum') if to_state else None
+        from_sha = from_state.get('gitshasum') if from_state else None
 
         if state in ('clean', 'added', 'modified', None):
             # assign present gitsha to any record
             # state==None can only happen for subdatasets that
             # already existed, so also assign a sha for them
-            props['gitshasum'] = to_sha
+            if to_sha:
+                # with a typechange there would be no gitsha
+                # for the new content, despite a known modification
+                props['gitshasum'] = to_sha
             if 'bytesize' in to_state:
                 # if we got this cheap, report it
                 props['bytesize'] = to_state['bytesize']
             elif state == 'clean' and 'bytesize' in from_state:
                 # no change, we can take this old size info
                 props['bytesize'] = from_state['bytesize']
+
         if state in ('clean', 'modified', 'deleted', None):
             # assign previous gitsha to any record
             # state==None can only happen for subdatasets that
             # already existed, so also assign a sha for them
             props['prev_gitshasum'] = from_sha
+
+        # current type reporting
+        if modified_in_worktree is not False:
+            # we have a report of a modified type, include if
+            # not None (i.e. vanished)
+            if modified_in_worktree is not None:
```

Comment:
This is probably the line that mode you think it should also take `None`, but I currently think this conditional is always true.

Reply:
then it better becomes an `assert modified_in_worktree is not None` ?